aerogear/aerogear-js-sdk
# Enable offline mutations when server is unreachable

### Description

- allow updates on items created while offline
  - when back online, operations in mutations queue are send one by one (waiting for previous one to finish). If there was optimistic id for the operation, and server returns real id, next operations with this optimistic id in queue are updated with this new id
- enable offline mutations when server is unreachable (e.g. down)
  - two more links were added to `compositeQueueLink`
    - `apollo-link-error` - so that offlineQueueLink can determine there was a network error
    - `apollo-link-retry` - this was added so that operation in offline queue is retried until there is no network error
  - for cache updates when replaying offline mutations, `proxyUpdate` function must be provided by user

https://issues.jboss.org/browse/AEROGEAR-8333
https://issues.jboss.org/browse/AEROGEAR-8334
https://issues.jboss.org/browse/AEROGEAR-8345

#### Verification steps

1. `npm link` those changes to `apollo-voyager-ionic-example` app
2. perform next steps in `apollo-voyager-ionic-example` app

For `allow updates on items created while offline` verification:

1. go offline
2. create item then make some updates on it
3. go back online
4. when offline updates are performed you should still see the updated item

For `enable offline mutations when server is unreachable` verification:

1. kill server
2. make some mutations
3. start the server again
4. offline mutations should be processed after short time

#### Checklist

- [x] update ionic example app - https://github.com/aerogear/apollo-voyager-ionic-example/pull/27
- [x] add verification steps
